llvm/include/llvm/SYCLLowerIR/ModuleSplitter.h

```diff
@@ -289,35 +291,41 @@ void dumpEntryPoints(const Module &M, bool OnlyKernelsAreEntryPoints = false,
                      const char *Msg = "", int Tab = 0);
 #endif // NDEBUG
 
-struct SplitModule {
+struct ProcessedModule {
   std::string ModuleFilePath;
   util::PropertySetRegistry Properties;
   std::string Symbols;
 
-  SplitModule() = default;
-  SplitModule(const SplitModule &) = default;
-  SplitModule &operator=(const SplitModule &) = default;
-  SplitModule(SplitModule &&) = default;
-  SplitModule &operator=(SplitModule &&) = default;
+  ProcessedModule() = default;
+  ProcessedModule(const ProcessedModule &) = default;
+  ProcessedModule &operator=(const ProcessedModule &) = default;
+  ProcessedModule(ProcessedModule &&) = default;
+  ProcessedModule &operator=(ProcessedModule &&) = default;
 
-  SplitModule(std::string_view File, util::PropertySetRegistry Properties,
-              std::string Symbols)
+  ProcessedModule(std::string_view File, util::PropertySetRegistry Properties,
+                  std::string Symbols)
       : ModuleFilePath(File), Properties(std::move(Properties)),
         Symbols(std::move(Symbols)) {}
 };
 
-struct ModuleSplitterSettings {
+struct ModuleProcessingSettings {
   IRSplitMode Mode;
   bool OutputAssembly = false; // Bitcode or LLVM IR.
   StringRef OutputPrefix;
 };
 
-/// Parses the output table file from sycl-post-link tool.
-Expected<std::vector<SplitModule>> parseSplitModulesFromFile(StringRef File);
+SmallString<64>
+convertProcessingSettingsToString(const ModuleProcessingSettings &S);
 
-/// Splits the given module \p M according to the given \p Settings.
-Expected<std::vector<SplitModule>>
-splitSYCLModule(std::unique_ptr<Module> M, ModuleSplitterSettings Settings);
+/// Parses the output table file from sycl-post-link tool.
+Expected<std::vector<ProcessedModule>>
+parseProcessedModulesFromFile(StringRef File);
```

Comment:
i thought we wanted to seperate module splitting out into a seperate operation, possibly so that we could hook it into any upstream splitting api, but it seems like here we are generalizing these APIs, at least in the name (also, the file name still has split it in which is a little confusing)

what's the long term plan/architecture?

thanks

Reply:
The following processing pipeline comes only from perspectives of RAM usage and compilation time.
Get new split module -> handle spec constants -> handle ESIMD part -> save bc files to disk -> run SPIRVTranslator.

It is known that keeping split modules in the RAM is not really possible. It leads to the question where we should perform the spec const and ESIMD parts? Before saving to disk or after?

We could do the following scheme:
Split modules -> save bc files to disk -> read them again, perform spec const, ESIMD processing -> save bc files to disk again -> run SPIRVTranslator.
This scheme entails twice disk ops compared to the previous one.

The original scheme takes its roots from the usage of SPIRVTranslator. In case of usage of SPIRV backend it would be much easier like the following:
Split modules and save bc files -> read bc file, perform spec const, ESIMD processing -> run SPIRV backend -> save spirv file to disk.
I expect to move to this scheme once SPIRV backend becomes the main path. However, I don't know when it will happen.

I see your concerns that you need some API for thin-LTO. I was considering to come up with a splitting Pass similar to one that AMD have in llvm-project.

While SPIRVTranslator is the main tool you still need to handle spec constants and ESIMD somewhere in thin-LTO. Most likely, `post-split` processing part might be extracted in order to be invoked from the library and thin-LTO framework.

Reply:
thanks for the explanation. actually for thin-lto we call sycl-post-link early (in -c phase) and run processing there. that is actually enough for everything besides spec constants (that we know of right now), so that one i will do as part of device link, so i don't think there is a need for an API to call all sycl-post-link post-split processing for thinlto, at least not at the moment